### PR TITLE
Create a shim for Spark 3.2.0 development

### LINKED ISSUE #1490: [FEA] Add Apache Spark 3.2.0 shim layer
**Is your feature request related to a problem? Please describe.**
Apache spark split the branch for the 3.1.X line which means master branch is now 3.2.0-SNAPSHOT. We need to add a shim layer for this.

### PR DESCRIPTION
**Pull request by gerashegalov**
Signed-off-by: Gera Shegalov <gera@apache.org>

Add a shim provider for Spark 3.2 development branch.

Splitting this from #1688 to reduce its scope

Closes #1490

### CODE REVIEW

File: integration_tests/run_pyspark_from_build.sh [hunk 1]
```diff
@@ -25,18 +25,10 @@ then
     >&2 echo "SPARK_HOME IS NOT SET CANNOT RUN PYTHON INTEGRATION TESTS..."
 else
     echo "WILL RUN TESTS WITH SPARK_HOME: ${SPARK_HOME}"
-    # support alternate local jars NOT building from the source code
```
Comment (jlowe) [inline integration_tests/run_pyspark_from_build.sh:28]: Why does this revert #1695?

File: integration_tests/pom.xml [hunk 2]
```diff
@@ -56,6 +56,12 @@
                 <spark.test.version>3.1.1-SNAPSHOT</spark.test.version>
             </properties>
         </profile>
+        <profile>
+            <id>spark320tests</id>
+            <properties>
+                <spark.test.version>3.2.0-SNAPSHOT</spark.test.version>
```
Comment (jlowe) [inline integration_tests/pom.xml:62]: Shouldn't this be using `${spark320.version}` from the parent POM?

File: pom.xml [hunk 3]
```diff
@@ -176,6 +176,7 @@
         <spark301.version>3.0.1</spark301.version>
         <spark302.version>3.0.2-SNAPSHOT</spark302.version>
         <spark311.version>3.1.1-SNAPSHOT</spark311.version>
+        <spark320.version>3.2.0-SNAPSHOT</spark320.version>
```
Comment (tgravescs) [inline pom.xml:179]: probably need to add something for the tests-spark310+/ tests.

**Review by tgravescs — COMMENTED** [review]
we also need to update the docs and create RapidsShuffleManager.scala for 3.2.0.

I've been creating a RapidsShuffleManager.scala so that its easy for the user to set it to match the version, see:
https://nvidia.github.io/spark-rapids/docs/get-started/getting-started-on-prem.html#enabling-rapidsshufflemanager
like:
Spark 3.1.0 (com.nvidia.spark.rapids.spark310.RapidsShuffleManager)

File: integration_tests/run_pyspark_from_build.sh (hunk 1 above)
Comment (gerashegalov) [reply integration_tests/run_pyspark_from_build.sh:28]: mistake when factoring this PR out of #1688, did not mean to include this file at all, will drop it

File: integration_tests/pom.xml (hunk 2 above)
Comment (gerashegalov) [reply integration_tests/pom.xml:62]: I had it initially like that but then decided to leave it consistent with the rest of the profiles in this pom.

Comment (jlowe) [reply integration_tests/pom.xml:62]: I think this (and the other profiles) should be fixed, otherwise updating the property on the command-line doesn't do the right thing here.

Comment (gerashegalov) [reply integration_tests/pom.xml:62]: Another change I've been eager to make is to replace most of the 2.12 mentions by `scala.binary.version`

Comment (jlowe) [reply integration_tests/pom.xml:62]: At one point we've replaced most of the ones that we can.  Maven squawked loudly when we tried putting the top-pom variable into the project names.  Maybe some more references have creeped in since I last looked at this.

Comment (gerashegalov) [reply integration_tests/pom.xml:62]: I think maven squawks only about using references inside `<parent>` :)

Comment (jlowe) [reply integration_tests/pom.xml:62]: No it wasn't that, obviously.  You can't expand a variable before it's defined.

**Comment by gerashegalov** [comment]
```bash
$ mvn verify -Pspark320tests -Dcuda.version=cuda11
$ find . -name scala-test-output.txt | xargs grep FAILED
./tests/target/surefire-reports/scala-test-output.txt:- IGNORE ORDER, WITH DECIMALS: short reduction aggs *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by string literal *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by float and string literal *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by floats *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by more_floats *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, NOT ON GPU[HashAggregateExec,AggregateExpression,AttributeReference,Alias,Literal,Min,Sum,Max,Average,Add,Multiply,Subtract,Cast,Count], WITH DECIMALS: partial on gpu: float basic aggregates group by more_floats *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, NOT ON GPU[HashAggregateExec,AggregateExpression,AttributeReference,Alias,Literal,Min,Sum,Max,Average,Add,Multiply,Subtract,Cast,Count,KnownFloatingPointNormalized,NormalizeNaNAndZero], WITH DECIMALS: final on gpu: float basic aggregates group by more_floats *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, WITH DECIMALS: nullable float basic aggregates group by more_floats *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, WITH DECIMALS: sum(floats) group by more_floats 2 partitions *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- INCOMPAT, IGNORE ORDER, WITH DECIMALS: empty df: reduction aggs *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- Test all supported casts with in-range values *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:- Join partitioned tables *** FAILED ***
./tests/target/surefire-reports/scala-test-output.txt:*** 12 TESTS FAILED ***
```

**Comment by gerashegalov** [comment]
build

**Comment by tgravescs** [comment]
so I assume by the above output that means tests are failing, did you look why? these pass for you with spark311 right?

File: shims/spark320/src/main/scala/com/nvidia/spark/rapids/shims/spark320/Spark320Shims.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+package com.nvidia.spark.rapids.shims.spark320
+
+import com.nvidia.spark.rapids.shims.spark311.Spark311Shims
+
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.adaptive.{BroadcastQueryStageExec, ShuffleQueryStageExec}
+import org.apache.spark.sql.execution.exchange.ReusedExchangeExec
+
+class Spark320Shims extends Spark311Shims {
+  override def reusedExchangeExecPfn: PartialFunction[SparkPlan, ReusedExchangeExec] = {
```
Comment (tgravescs) [inline shims/spark320/src/main/scala/com/nvidia/spark/rapids/shims/spark320/Spark320Shims.scala:26]: what changed here to cause this?  Can we add a comment

File: shims/spark320/src/main/scala/com/nvidia/spark/rapids/spark320/RapidsShuffleManager.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
Comment (tgravescs) [inline shims/spark320/src/main/scala/com/nvidia/spark/rapids/spark320/RapidsShuffleManager.scala:2]: remove 2020

**Comment by gerashegalov** [comment]
> so I assume by the above output that means tests are failing, did you look why? these pass for you with spark311 right?

@tgravescs yes, `-Pspark311tests` results are green :heavy_check_mark: . I am going to look at the `-Pspark320tests failures

**Comment by gerashegalov** [comment]
**Status**

:heavy_check_mark:  FIXED `ClassNotFoundException: org.apache.spark.sql.catalyst.errors.package$`

:x:  `IGNORE ORDER, WITH DECIMALS: short reduction aggs *** FAILED ***`
```
  Running on the GPU and on the CPU did not match 
  CPU: WrappedArray(WrappedArray(163835, 114683, 8, 14335.375, 15.0))
  
  GPU: WrappedArray(WrappedArray(163835, -16389, 8, 14335.375, 15.0)) (SparkQueryCompareTestSuite.scala:708)
```

:x:  `INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by string literal *** FAILED ***`
```
Cause: java.lang.AssertionError: Type conversion is not allowed from Table{columns=[ColumnVector{rows=1, type=INT32, nullCount=Optional.empty, offHeap=(ID: 22210 7f2590476f00)}, ColumnVector{rows=1, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22211 7f259047c420)}, ColumnVector{rows=1, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22212 7f2590476d90)}, ColumnVector{rows=1, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22213 7f259047d8d0)}, ColumnVector{rows=1, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22214 7f259047f3d0)}, ColumnVector{rows=1, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22215 7f259047c0d0)}, ColumnVector{rows=1, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22216 7f2590478ea0)}, ColumnVector{rows=1, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22217 7f2590480030)}, ColumnVector{rows=1, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22218 7f259047d6b0)}, ColumnVector{rows=1, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22219 7f2590475d20)}, ColumnVector{rows=1, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22220 7f2590475e30)}, ColumnVector{rows=1, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22221 7f2590478b90)}], cudfTable=139799310887712, rows=1} to [IntegerType, FloatType, DoubleType, FloatType, FloatType, FloatType, FloatType, DoubleType, DoubleType, DoubleType, LongType, LongType][0m
[31m  at com.nvidia.spark.rapids.GpuColumnVectorFromBuffer.from(GpuColumnVectorFromBuffer.java:69)
```

:x:  `INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by float and string literal`
```
Cause: java.lang.AssertionError: Type conversion is not allowed from Table{columns=[ColumnVector{rows=7, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22310 7f25900134b0)}, ColumnVector{rows=7, type=STRING, nullCount=Optional.empty, offHeap=(ID: 22311 7f2590016d80)}, ColumnVector{rows=7, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22312 7f2590474380)}, ColumnVector{rows=7, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22313 7f259046ba30)}, ColumnVector{rows=7, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22314 7f2590014890)}, ColumnVector{rows=7, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22315 7f25900147f0)}, ColumnVector{rows=7, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22316 7f25904745c0)}, ColumnVector{rows=7, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22317 7f2590016680)}, ColumnVector{rows=7, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22318 7f2590482e50)}, ColumnVector{rows=7, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22319 7f2590482ef0)}, ColumnVector{rows=7, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22320 7f2590482f90)}, ColumnVector{rows=7, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22321 7f2590483030)}, ColumnVector{rows=7, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22322 7f25904830d0)}], cudfTable=139799309909712, rows=7} to [FloatType, StringType, FloatType, DoubleType, FloatType, FloatType, FloatType, FloatType, DoubleType, DoubleType, DoubleType, LongType, LongType][0m
```

:x: `INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by floats *** FAILED ***[0m`
```
Cause: java.lang.AssertionError: Type conversion is not allowed from Table{columns=[ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22478 7f2590013a80)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22479 7f2590287a90)}, ColumnVector{rows=6, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22480 7f25902fa170)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22481 7f25902b8a70)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22482 7f2590307800)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22483 7f25902f1fc0)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22484 7f25902f1f70)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22485 7f25902f1e90)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22486 7f25902ef040)}, ColumnVector{rows=6, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22487 7f2590491aa0)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22488 7f25902f1b70)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22489 7f25902f1c10)}], cudfTable=139799309532224, rows=6} to [FloatType, FloatType, DoubleType, FloatType, FloatType, FloatType, FloatType, DoubleType, DoubleType, DoubleType, LongType, LongType]
```

:x: `INCOMPAT, IGNORE ORDER, WITH DECIMALS: float basic aggregates group by more_floats *** FAILED ***
```
Cause: java.lang.AssertionError: Type conversion is not allowed from Table{columns=[ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22571 7f2590012c00)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22572 7f25901f41e0)}, ColumnVector{rows=6, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22573 7f25901f3fd0)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22574 7f25904986d0)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22575 7f2590498380)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22576 7f2590491750)}, ColumnVector{rows=6, type=FLOAT32, nullCount=Optional.empty, offHeap=(ID: 22577 7f2590491700)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22578 7f259049c170)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22579 7f25902f1df0)}, ColumnVector{rows=6, type=FLOAT64, nullCount=Optional.empty, offHeap=(ID: 22580 7f25902b8fc0)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22581 7f259049be50)}, ColumnVector{rows=6, type=INT64, nullCount=Optional.empty, offHeap=(ID: 22582 7f259049bef0)}], cudfTable=139799311269296, rows=6} to [FloatType, FloatType, DoubleType, FloatType, FloatType, FloatType, FloatType, DoubleType, DoubleType, DoubleType, LongType, LongType]
```
```

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 6]
```diff
@@ -1857,7 +1857,7 @@ object GpuOverrides {
           }
         }
 
-        override def convertToGpu(child: Expression): GpuExpression = GpuSum(child)
+        override def convertToGpu(child: Expression): GpuExpression = GpuSum(child, a.dataType)
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:1860]: this kind of change will be needed for avg for decimal support, the current tests are passing. Easiest way to identify expressions requiring this treatment in Spark:
```
➜  spark git:(master) ag 'def dataType: DataType = resultType'
sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala
47:  override def dataType: DataType = resultType

sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala
51:  override def dataType: DataType = resultType
```

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java [hunk 7]
```diff
@@ -107,44 +107,61 @@ public static synchronized void debug(String name, HostColumnVector hostCol) {
...
   }
 
+  private static void debugInteger(HostColumnVector hostCol, DType intType) {
+    for (int i = 0; i < hostCol.getRowCount(); i++) {
+      if (hostCol.isNull(i)) {
+        System.err.println(i + " NULL");
+      } else {
+        String value;
+        switch (intType.getTypeId()) {
+          case INT8:
```
Comment (jlowe) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:140]: Given this is only supposed to be called when the caller already knows the type is an integer, I think we can avoid redundantly switching on every possible enum and instead leverage `DType#getSizeInBytes` to know whether we need to get a byte, short, int, or long.

File: shims/spark320/pom.xml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+          </resource>
+          <resource>
+            <directory>src/main/resources</directory>
+          </resource>
+        </resources>
+    </build>
+
+    <dependencies>
+	<dependency>
+            <groupId>com.nvidia</groupId>
```
Comment (jlowe) [inline shims/spark320/pom.xml:80]: Nit: The indentation in this file is not consistent.  Sometimes it's 2, sometimes it's 4.  My preference is 2 to match the code, but I don't care to much as long as it's consistently applied.

File: shims/spark320/pom.xml [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            <artifactId>rapids-4-spark-shims-spark311_${scala.binary.version}</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-sql_${scala.binary.version}</artifactId>
+            <version>${spark320.version}</version>
+	    <scope>provided</scope>
```
Comment (jlowe) [inline shims/spark320/pom.xml:88]: Indentation inconsistency within the same block

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/SparkShimServiceProvider.scala [hunk 10]
```diff
@@ -22,7 +22,8 @@ object SparkShimServiceProvider {
   // keep 3.1.0 snapshot version for now until 3.1.1 is released
   val VERSION310 = SparkShimVersion(3, 1, 0)
   val VERSION = SparkShimVersion(3, 1, 1)
-  val VERSIONNAMES = Seq(s"$VERSION310-SNAPSHOT", s"$VERSION", s"$VERSION-SNAPSHOT")
+  val VERSIONNAMES: Seq[String] = Seq(VERSION310, VERSION)
+    .flatMap(v => Seq(s"$v", s"$v-SNAPSHOT"))
```
Comment (jlowe) [inline shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/SparkShimServiceProvider.scala:26]: Please revert this change.  It's going to conflict with the change coming in #1832 and isn't necessary for this PR.

File: shims/spark320/src/main/scala/com/nvidia/spark/rapids/shims/spark320/Spark320Shims.scala [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.apache.spark.sql.catalyst.trees.TreeNode
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.adaptive.{BroadcastQueryStageExec, ShuffleQueryStageExec}
+import org.apache.spark.sql.execution.exchange.ReusedExchangeExec
+
+class Spark320Shims extends Spark311Shims {
+
+  override def getSparkShimVersion: ShimVersion = SparkShimServiceProvider.VERSION320
+
```
Comment (jlowe) [inline shims/spark320/src/main/scala/com/nvidia/spark/rapids/shims/spark320/Spark320Shims.scala:30]: I think this needs the following override:
```
  override def getRapidsShuffleManagerClass: String = {
    classOf[RapidsShuffleManager].getCanonicalName
  }
```

**Review by jlowe — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 12]
```diff
@@ -140,6 +140,12 @@
                 <module>tests-spark310+</module>
             </modules>
         </profile>
+        <profile>
+            <id>spark320tests</id>
+            <modules>
+                <module>tests-spark310+</module>
```
Comment (tgravescs) [inline pom.xml:146]: so I think we need more changes in the tests-spark310+/pom.xml file to change the spark version based on the profile. Otherwise it will always use spark 3.1.0

File: shims/spark320/src/main/scala/com/nvidia/spark/rapids/shims/spark320/Spark320Shims.scala (hunk 11 above)
Comment (gerashegalov) [reply shims/spark320/src/main/scala/com/nvidia/spark/rapids/shims/spark320/Spark320Shims.scala:31]: thanks for catching this. added with a unit test

File: pom.xml (hunk 12 above)
Comment (gerashegalov) [reply pom.xml:168]: thanks for catching this. i consolidated version profiles into parent pom, I think it would be simpler. Not a maven expert but it seems to work as expected.

**Review by jlowe — APPROVED** [review]

**Comment by gerashegalov** [comment]
build

**Review by tgravescs — APPROVED** [review]
